### PR TITLE
Vector Quantized Embeddings

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Vector quantization is an old method introduced with the VQ-VAE (https://arxiv.org/abs/1711.00937) for achieving high fidelity image generation. It still remains a core component in many landmark models: 

- the DALL-E models use it for image generation by autoregressively predicting image token IDs
- Latent diffusion uses a VAE (not sure if it was VQ or continuous) to perform the diffusion process in the latent space
- Muse by Google uses VQVAE similarly for image generation
- Make-A-Video by Meta uses it for video generation from text

To continue to support image and more modalities in the future, we need to add the core VQ embedding layer as a module. This module will maintain the codebook embeddings and perform the lookup for discretizing an encoder input.

I've adapted the `Codebook` layer from torchmultimodal (found [here](https://github.com/facebookresearch/multimodal/blob/main/torchmultimodal/modules/layers/codebook.py#L31), apparently it shares authors with the author of this PR...). I've removed features such as initializing embeddings with encoder input for faster convergence to help simplify the code. I've kept the core embedding lookup functionality and the EMA update so that it could still be trained if users want to experiment.

#### Changelog
What are the changes made in this PR?
* Add `VectorQuantizedEmbeddings` and requisite tests

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/modules/vq_embeddings.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+class VectorQuantizedEmbeddings(nn.Module):
+    """
+    Vector quantized embedding layer that takes in the output of an encoder
+    and performs a nearest-neighbor lookup in the embedding space.
+    Vector quantization was introduced in Oord et al. 2017 (https://arxiv.org/pdf/1711.00937.pdf)
+    to generate high-fidelity images, videos, and audio data.
+    The embedding weights are trained with exponential moving average updates as described
+    in original paper.
+
+    Code was adapted from torchmultimodal's `Codebook module
```
Comment (pbontrager) [inline torchtune/modules/vq_embeddings.py:23]: I was going to comment that I'm more familiar with this being called a codebook. Any reason for not calling it that here? Or VQCodebook?

File: torchtune/modules/vq_embeddings.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        learnable: bool = False,
+        epsilon: float = 1e-7,
+    ) -> None:
+        super().__init__()
+        # Embedding weights and parameters for EMA update will be registered to buffer, as they
+        # will not be updated by the optimizer but are still model parameters.
+        # code_usage and code_avg correspond with N and m, respectively, from Oord et al.
+        randn_init_embedding = torch.randn(num_embeddings, embedding_dim)
+        self.register_buffer("embedding", randn_init_embedding.clone())
+        if learnable:
```
Comment (pbontrager) [inline torchtune/modules/vq_embeddings.py:54]: I don't really like this variable because it could disagree with the nn.module.training attribute. Is this only needed for EMA? Why wouldn't the ema code handle this?

File: torchtune/modules/vq_embeddings.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        token_ids = token_ids_flat.view(bsz, seq_len)
+
+        return quantized, token_ids
+
+    def extra_repr(self) -> str:
+        return "num_embeddings={}, embedding_dim={}".format(
+            self.num_embeddings, self.embedding_dim
+        )
+
+    def lookup(self, token_ids: Tensor) -> Tensor:
```
Comment (pbontrager) [inline torchtune/modules/vq_embeddings.py:171]: nit: do we want to use "lookup" here or something more consistent with the rest of the library like deocde?

**Review by pbontrager — COMMENTED** [review]
This looks like a nice implementation. I left some nit comments and questions. But my larger question is if all of this is needed in torchtune? How much of this code is just needed to support training a codebook? Do we want/need that or should we just house the minimal amount needed for inference?

**Comment by RdoubleA** [comment]
> Do we want/need that or should we just house the minimal amount needed for inference?

Open to either, was thinking since I'm putting this in modules it should at least be trainable. But it would simplify the class significantly if we only supported inference for this.

File: torchtune/modules/vq_embeddings.py (hunk 1 above)
Comment (RdoubleA) [reply torchtune/modules/vq_embeddings.py:23]: no strong opinions, I've seen it called both. we could do `VectorQuantizedCodebook`?

**Comment by pbontrager** [comment]
It's up to you here. It could be nice to already have training logic, but I'm not sure if we know what a training recipe for encoders would look like for us. If we decide to support training an encoder in the future we might make different decisions on how to split the loss logic from the module logic. If we wanted to make this module finetunable with the rest of the LLM, would we use these same methods or allow full override of the codebook?

File: torchtune/modules/vq_embeddings.py (hunk 1 above)
Comment (pbontrager) [reply torchtune/modules/vq_embeddings.py:23]: I guess that could be confusing too since it's just a "normal" codebook.  Maybe AutoEncoderCodebook? Or just leave it.

Comment (RdoubleA) [reply torchtune/modules/vq_embeddings.py:23]: Having `VectorQuantized` is the most descriptive imo, so I will leave it as it

**Review by felipemello1 — APPROVED** [review]
